tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,357 @@
+/**
+ * @file   mem_filesystem.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2020 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements the in-memory filesystem class
+ */
+
+#include <sstream>
+#include <unordered_set>
+
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/filesystem/mem_filesystem.h"
+#include "tiledb/sm/global_state/global_state.h"
+#include "tiledb/sm/misc/utils.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+class MemFilesystem::FSNode {
+ public:
+  FSNode(const std::string& name)
+      : name_(name){};
+
+  virtual ~FSNode() = default;
+
+  std::string get_name() const {
+    return name_;
+  }
+
+  virtual bool is_dir() const = 0;
+  virtual bool is_file() const = 0;
+  virtual std::vector<std::string> ls(const std::string& dir) const = 0;
+  virtual bool has_child(const std::string& child) const = 0;
+
+  // check if map instead
+  std::unordered_map<std::string, FSNode*> children_;
```

Comment:
Unless we really need this, I think a std::vector<FSNode> would be sufficient. I don't think the `FSNode` need to be pointers because they shouldn't be shared. Having the map here imposes an invariant that the key must match the object's `name_`.

Reply:
The only reason for using an unordered_map is the ease/performance of navigating through the tree to get/remove a file. I also thought of using a simple map that will keep the contents sorted by filename, which is something we could benefit from (in the expense of performance of course). 
If you still prefer a vector better it's fine with me, we will just need to walk each time all the directory contents to find the next directory in a given path. I will let you decide :) I am not sure I understand why it's inconvenient to store nodes by name_ though?
Regarding pointers, see my reply to your other related comment.

Reply:
The inconvenience is that it introduces redundant state that may make operations more complex. Consider implementing a rename with a vector vs a map:

// vector
1. Locate the child to rename.
2. Lock on the child.
3. Rename the child.
4. Unlock on the child.

// map
1. Locate the FSNode parent.
2. Lock the parent.
3. Update the key in the map.
4. Lock the child.
5. Rename the child.
6. Unlock the child.
7. Unlock the parent.

I am totally fine with this, I was just pointing out that we don't necessarily need it in a first implementation (hoping that it may simplify your patch). Let's keep it.